File: leetcode/079-word-search.py

def does_exist(board, word, word_idx, i, j, visited):
    if len(word) == word_idx:
        return True
    
    for (di, dj) in ((-1, 0), (1, 0), (0, -1), (0, 1)):
        ci = i + di
        cj = j + dj
        if not 0 <= ci < len(board) or not 0 <= cj <= len(board[ci]) \
            or board[ci][cj] != word[word_idx] or (ci, cj) in visited:
            continue
        visited.add((ci, cj))
        has_path = does_exist(
            board,
            word,
            word_idx + 1,
            ci,
            cj,
            visited
        )
        visited.remove((ci, cj))
        if has_path:
            return True
    return False

class Solution(object):
    def exist(self, board, word):
        for i, row in enumerate(board):
            for j, val in enumerate(row):
                if val == word[0] and \
                    does_exist(board, word, 1, i, j, {(i, j)}):
                    return True
        return False

File: leetcode/079-word-search.js
var exist = function(board, word) {
    if (!board || !word) {
        return false;
    }

    let foundIt = false;

    const dfs = function(row, col, markedGrid, wordSoFar) {
        // console.log("markedGrid: ", markedGrid);
        if (foundIt) {
            return
        }
        const newMarkedGrid = markedGrid.map(row => row.slice())
        const letter = newMarkedGrid[row][col];
        // console.log("row: ", row, "; col: ", col, "word:  ", wordSoFar, "; letter: ", letter)
        // console.log("grid: ", markedGrid)
        if (!letter) {
            return
        }
        let nextWord = wordSoFar + letter;
        if (word === nextWord) {
            foundIt = true;
            // console.log("Found it!!")
            return;
        } else if (word.indexOf(nextWord) === 0) {
            // keep going
            markedGrid[row][col] = null;
            // up
            if (row - 1 >= 0) {
                dfs(row - 1, col, markedGrid.map(row => row.slice()), nextWord);
            }

            // down
            if (row + 1 < markedGrid.length) {
                dfs(row + 1, col, markedGrid.map(row => row.slice()), nextWord)
            }

            // left
            if (col - 1 >= 0) {
                dfs(row, col - 1, markedGrid.map(row => row.slice()), nextWord)
            }

            // right
            if (col + 1 < markedGrid[row].length) {
                dfs(row, col + 1, markedGrid.map(row => row.slice()), nextWord)
            }
        }
    }

    for (var row = 0; row < board.length; row++) {
        for (var col = 0; col < board[row].length; col++) {
            const letter = board[row][col]
            if (letter === word.charAt(0)) {
                const grid = board.map(row => row.slice())
                dfs(row, col, grid, '')
                if (foundIt) {
                    return true
                }
            }
        }
    }

    return false
}

// console.log(exist([
//     ['C','A','A'],
//     ['A','A','A'],
//     ['B','C','D']
//   ],"AAB")) // true

  console.log(exist([
    ['A','B','C','E'],
    ['S','F','E','S'],
    ['A','D','E','E']
  ],"ABCESEEEFS")) // true


// console.log(exist([
//     ['A','B','C','E'],
//     ['S','F','C','S'],
//     ['A','D','E','E']
//   ],"ABCCED")) // true
// console.log(exist([
//     ['A','B','C','E'],
//     ['S','F','C','S'],
//     ['A','D','E','E']
//   ],"SEE")) // true
// console.log(exist([
//     ['A','B','C','E'],
//     ['S','F','C','S'],
//     ['A','D','E','E']
//   ],"ABCB")) // true